README.md
yii2-sitemap-generator
===================

Yii2 component for generation sitemap.xml

It is reworked package, based on [Dreamjobs yii2 sitemap generator](https://github.com/dreamjobs/yii2-sitemap-generator)

Installation
------------

Run command

`composer require itstructure/yii2-sitemap-generator "~1.0.0"`

Configuration
-------------

Set in `"./console/config/main.php"`

```php
return [
    'controllerMap' => [
        'sitemap' => [
            'class' => 'Itstructure\Sitemap\SitemapController',
            'baseUrl' => 'https://example.com',
            'modelsPath' => '@console/models/sitemap', // Sitemap-data models directory
            'modelsNamespace' => 'console\models\sitemap', // Namespace in [[modelsPath]] files
            'savePathAlias' => '@frontend/web', // Where would be placed the generated sitemap-files
            'sitemapFileName' => 'sitemap.xml', // Name of main sitemap-file in [[savePathAlias]] directory
        ],
    ],
];
```

Set in `"./environments/prod/console/config/main-local.php"`

```php
'components' => [
    // fix console create url
    'urlManager' => [
        'baseUrl' => 'http://example.com',
    ],
],
```

Set in `"./environments/dev/console/config/main-local.php"`

```php
'components' => [
    // fix console create url
    'urlManager' => [
        'baseUrl' => 'http://example.local',
    ],
],
```

Run command

```code
php ./init
```

OR just apply same config to your "/console/config/main-local.php"

ALSO you can merge urlManager rules from frontend(or common) to console config.
Just change "/console/config/main.php" file:

```php
// get config of urlManager from frontend for correctly create urls in console app
$frontend = require(__DIR__ . '/../../frontend/config/main.php');
$frontendUrlManager = [
    'components' => [
        'urlManager' => $frontend['components']['urlManager'],
    ],
];

// ...

// Merge frontend urlManager config with console application main config
return yii\helpers\ArrayHelper::merge($frontendUrlManager, [
    'id' => 'app-console',
    // ...
];
```

Also useful append .gitignore for ignore all generated sitemaps files:

```code
# sitemaps
/frontend/web/sitemap*.xml
```

TBD: creating sitemap-data models

Usage
-----
Run Yii console command in project root:

```code
php ./yii sitemap
```


changelog.md
### CHANGE LOG:

**1.0.0 July 06, 2020:**
- Create component with the first release.


SitemapBuilder.php
<?php

namespace Itstructure\Sitemap;

use Yii;
use yii\base\BaseObject;
use yii\console\Exception;

/**
 * Class SitemapBuilder
 * @package Itstructure\Sitemap
 */
class SitemapBuilder extends BaseObject
{
    /** @var string Path to saving sitemap-files. As webroot: "http://example.com" */
    public $savePathAlias = '@frontend/web';

    /** @var string Name of sitemap-file, saved to [[savePathAlias]]: "sitemap.xml" */
    public $sitemapFileName = 'sitemap.xml';

    /** @var array List of sitemap schemas */
    public $schemas = [
        'xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"',
    ];

    /** @var int Max count of <url> per one file */
    public $urlsPerFile = 10000;

    /** @var int Count items in current file */
    private $_itemsCount = 0;

    /** @var resource File pointer */
    private $_file;

    /** @var array List of xml-files for sitemap index */
    private $_filesList = [];

    /**
     * @var null
     */
    public $baseUrl = null;

    public function writeHeader()
    {
        $header = '<?xml version="1.0" encoding="UTF-8"?>' . PHP_EOL;
        $header .= '<urlset ' . implode(PHP_EOL . "\t", $this->schemas) . '>' . PHP_EOL;

        $this->appendToFile($header);
    }

    /**
     * @return bool
     */
    public function writeFooter()
    {
        $footer = '</urlset>';

        $this->appendToFile($footer);

        // Close current file
        return @fclose($this->_file);
    }

    /**
     * Append to opened file new content
     * @param string $content
     * @throws \yii\console\Exception
     * @return int
     */
    public function appendToFile($content)
    {
        $result = @fwrite($this->_file, $content);

        if ($result === false) {
            throw new Exception("I'm can not wrtite to file");
        }

        return $result;
    }

    /**
     * Checks file pointer
     * @return bool
     */
    public function isFileClosed()
    {
        return !is_resource($this->_file);
    }

    /**
     * Create/rewrite new file with increment number
     * @return resource Pointer of created file
     * @throws \yii\console\Exception
     */
    public function beginNewFile()
    {
        $nameParts = explode('.', $this->sitemapFileName);
        $ext = array_pop($nameParts);
        $filename = implode('.', $nameParts) . (count($this->_filesList) + 1) . '.' . $ext;

        $fullPath = Yii::getAlias($this->savePathAlias) . DIRECTORY_SEPARATOR . $filename;

        $f = @fopen($fullPath, 'w');

        if ($f === false) {
            throw new Exception("I'm cannot create file '$fullPath'");
        }

        // Add new filename to list of exists
        $this->_filesList[] = $filename;

        $this->_file = $f;

        // Write sitemap header
        $this->writeHeader();
        // Reset count of items
        $this->_itemsCount = 0;

        return $this->_file;
    }

    /**
     * Create new url node object
     * @param array $config Object config
     * @return SitemapUrlNode
     */
    public function newUrl($config = [])
    {
        return Yii::createObject(array_merge([
            'class' => SitemapUrlNode::className(),
        ]), $config);
    }

    /**
     * Append new <url> section to current file
     * @param SitemapUrlNode $url
     * @return bool
     */
    public function writeUrl(SitemapUrlNode $url)
    {
        // if count of written items more than allowed per one xml sitemap file
        if (++$this->_itemsCount > $this->urlsPerFile) {
            // Ending of current file
            $this->writeFooter();

            // Starting new file
            $this->beginNewFile();

            return $this->writeUrl($url);
        }

        // Write new <url> section to current file
        return (bool)$this->appendToFile($url . PHP_EOL);
    }

    public function start()
    {
        $this->beginNewFile();
    }

    /**
     * @return bool
     */
    public function finish()
    {
        // complete file if needed
        if (!$this->isFileClosed()) {
            $this->writeFooter();
        }

        $savePath = Yii::getAlias($this->savePathAlias);
        $sitemapFilename = $savePath . DIRECTORY_SEPARATOR . $this->sitemapFileName;

        if (count($this->_filesList) === 1) {
            // rename created file to actual sitemap name
            $oldName = $savePath . DIRECTORY_SEPARATOR . $this->_filesList[0];

            if (!rename($oldName, $sitemapFilename)) {
                throw new Exception("I'm cannot rename file from '$oldName' to '$sitemapFilename'");
            }

            return true;
        }

        // if count of files greater than 1 - create sitemap-index file

        $this->_file = @fopen($sitemapFilename, 'w');

        $xml = [];
        $xml[] = '<?xml version="1.0" encoding="UTF-8"?>';
        $xml[] = '<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">';

        foreach ($this->_filesList as $file) {
            $xml[] = "\t<sitemap>";
            $xml[] = "\t\t<loc>" . $this->baseUrl . '/' . $file . '</loc>';
            // @todo set actual lastmod value
            $xml[] = "\t\t<lastmod>" . date(DATE_W3C) . '</lastmod>';
            $xml[] = "\t</sitemap>";
        }

        $xml[] = "</sitemapindex>";

        $this->appendToFile(implode(PHP_EOL, $xml));

        // Close current file
        return @fclose($this->_file);
    }
}


SitemapUrlNode.php
<?php

namespace Itstructure\Sitemap;

use yii\base\BaseObject;
use Itstructure\Sitemap\interfaces\Basic;

/**
 * Class SitemapUrlNode
 * @package Itstructure\Sitemap
 */
class SitemapUrlNode extends BaseObject
{
    public $loc;
    public $lastmod;
    public $changefreq;
    public $priority;
    public $images = [];
    public $alternateLinks = [];

    /**
     * SitemapUrlNode constructor.
     * @param array $config
     */
    public function __construct($config = [])
    {
        parent::__construct($config);
    }

    /**
     * Convert this url node object to valid xml-string
     * @return string
     */
    public function __toString()
    {
        if (empty($this->loc)) {
            // Empty loc is not allowed in sitemap.xml
            return '';
        }

        $url = ['<url>'];

        $url[] = "\t<loc>" . static::prepareUrl($this->loc) . '</loc>';

        // Basic data
        if ($this->lastmod !== null) {
            $lastmod = is_int($this->lastmod) ? $this->lastmod : strtotime($this->lastmod);
            $url[] = "\t<lastmod>" . date(DATE_W3C, $lastmod) . '</lastmod>';
        }

        if ($this->changefreq !== null) {
            $url[] = "\t<changefreq>{$this->changefreq}</changefreq>";
        }

        if ($this->priority !== null) {
            $url[] = "\t<priority>{$this->priority}</priority>";
        }

        // Google images data
        foreach ($this->images as $image) {
            $url[] = "\t<image:image>";
            $url[] = "\t\t<image:loc>" . static::prepareUrl($image['loc']) . '</image:loc>';

            if ($image['caption'] !== null) {
                $url[] = "\t\t<image:caption>" . $image['caption'] . '</image:caption>';
            }
            if ($image['geoLocation'] !== null) {
                $url[] = "\t\t<image:geo_location>" . $image['geoLocation'] . '</image:geo_location>';
            }
            if ($image['title'] !== null) {
                $url[] = "\t\t<image:title>" . $image['title'] . '</image:title>';
            }
            if ($image['license'] !== null) {
                $url[] = "\t\t<image:license>" . $image['license'] . '</image:license>';
            }

            $url[] = "\t</image:image>";
        }

        // Google alternate hreflang data
        foreach ($this->alternateLinks as $hreflang => $href) {
            $url[] = "\t" . '<xhtml:link rel="alternate" hreflang="' . $hreflang . '" ' . 'href="' .
                static::prepareUrl($href) . '" />';
        }

        $url[] = '</url>';

        return implode(PHP_EOL, $url);
    }

    /**
     * Set location value(url)
     * @param string $loc
     * @return $this
     */
    public function loc($loc)
    {
        $this->loc = $loc;

        return $this;
    }

    /**
     * Set last modificated time value
     * @param string $lastmod String applicable to strtotime() function
     * @return $this
     */
    public function lastmod($lastmod)
    {
        $this->lastmod = $lastmod;

        return $this;
    }

    /**
     * Set the regularity of content changing for [[loc]]
     * @param string $changefreq
     * @return $this
     */
    public function changefreq($changefreq)
    {
        $this->changefreq = $changefreq;

        return $this;
    }

    /**
     * Set priority value
     * May be between 0.0 - 1.0
     * @param string $priority
     * @return $this
     */
    public function priority($priority = Basic::PRIORITY_5)
    {
        $this->priority = $priority;

        return $this;
    }

    /**
     * Add image to [[images]] set
     * @param string $loc
     * @param string|null $geoLocation
     * @param string|null $caption
     * @param string|null $title
     * @param string|null $license
     * @return $this
     */
    public function addImage($loc, $geoLocation = null, $caption = null, $title = null, $license = null)
    {
        if (empty($loc)) {
            return $this;
        }

        $image = [
            'loc' => $loc,
            'geoLocation' => $geoLocation,
            'caption' => $caption,
            'title' => $title,
            'license' => $license,
        ];
        $this->images[] = $image;

        return $this;
    }

    /**
     * Add alternate link to [[alternateLinks]] set
     * @param string $hreflang Language code
     * @param string $href     Absolute url to view material on $hreflang language
     * @return $this
     */
    public function addAlternateLink($hreflang, $href)
    {
        $this->alternateLinks[$hreflang] = $href;

        return $this;
    }

    /**
     * Prepare url to place in <loc> tag
     * @param string $url original url
     * @return string
     */
    public static function prepareUrl($url)
    {
        // $url = urlencode($url);

        $replacement = [
            '&' => '&amp;',
            "'" => '&apos;',
            '"' => '&quot;',
            '>' => '&gt;',
            '<' => '&lt;',
        ];

        $url = str_replace(array_keys($replacement), array_values($replacement), $url);

        return $url;
    }
}


SitemapArrayItem.php
<?php

namespace Itstructure\Sitemap;

/**
 * Class SitemapArrayItem
 * @package Itstructure\Sitemap
 */
class SitemapArrayItem
{
    /**
     * @var array
     */
    public $item = [];

    /**
     * SitemapArrayItem constructor.
     * @param array $item
     */
    public function __construct(array $item)
    {
        $this->item = $item;
    }

    /**
     * @param $name
     * @param null $default
     * @return mixed|null
     */
    public function getItemParam($name, $default = null)
    {
        return isset($this->item[$name]) ? $this->item[$name] : $default;
    }

    /**
     * @return mixed|null
     */
    public function getSitemapLoc()
    {
        return $this->getItemParam('loc');
    }

    /**
     * @return mixed|null
     */
    public function getSitemapLastmod()
    {
        return $this->getItemParam('lastmod');
    }

    /**
     * @return mixed|null
     */
    public function getSitemapChangefreq()
    {
        return $this->getItemParam('changefreq');
    }

    /**
     * @return mixed|null
     */
    public function getSitemapPriority()
    {
        return $this->getItemParam('priority');
    }

    /**
     * @return mixed|null
     */
    public function getSitemapAlternateLinks()
    {
        return $this->getItemParam('alternateLinks', []);
    }
} 

interfaces/Basic.php
<?php

namespace Itstructure\Sitemap\interfaces;

/**
 * Basic interface for general sitemap.xml file
 *
 * @property array $sitemapLanguages     List of app languages. Default is [Yii::$app->language].
 * @property int $sitemapBatchSize       Size of one part selected from DB. Default is 10.
 * @property bool $sitemapSwithLanguages If TRUE - DataHandler will be swith App language to currently handled language.
 *
 * @package Itstructure\Sitemap\interfaces
 */
interface Basic
{
    const CHANGEFREQ_ALWAYS = 'always';
    const CHANGEFREQ_HOURLY = 'hourly';
    const CHANGEFREQ_DAILY = 'daily';
    const CHANGEFREQ_WEEKLY = 'weekly';
    const CHANGEFREQ_MONTHLY = 'monthly';
    const CHANGEFREQ_YEARLY = 'yearly';
    const CHANGEFREQ_NEVER = 'never';

    const PRIORITY_0 = '0.0';
    const PRIORITY_1 = '0.1';
    const PRIORITY_2 = '0.2';
    const PRIORITY_3 = '0.3';
    const PRIORITY_4 = '0.4';
    const PRIORITY_5 = '0.5';
    const PRIORITY_6 = '0.6';
    const PRIORITY_7 = '0.7';
    const PRIORITY_8 = '0.8';
    const PRIORITY_9 = '0.9';
    const PRIORITY_10 = '1.0';

    /**
     * Get list of static content items on the specified $lang
     * @param string|null $lang Language code
     * @return array
     */
    public function getSitemapItems($lang = null);

    /**
     * Get [[ActiveQuery]] for searh models the specified $lang
     * @param string|null $lang Language code for searching models
     * @return \yii\db\ActiveQuery
     */
    public function getSitemapItemsQuery($lang = null);

    /**
     * [REQIRED] URL of the page. This URL must begin with the protocol (such as http) and end with a trailing slash,
     * if your web server requires it. This value must be less than 2,048 characters.
     * @url http://www.sitemaps.org/protocol.html#locdef
     * @param string|null $lang language of items content
     * @return string
     */
    public function getSitemapLoc($lang = null);

    /**
     * [OPTIONAL] The date of last modification of the file. This date should be in W3C Datetime format.
     * This format allows you to omit the time portion, if desired, and use YYYY-MM-DD.
     * Note that this tag is separate from the If-Modified-Since (304) header the server can return,
     * and search engines may use the information from both sources differently.
     * @url http://www.sitemaps.org/protocol.html#lastmoddef
     * @param string|null $lang language of items content
     * @return string
     */
    public function getSitemapLastmod($lang = null);

    /**
     * [OPTIONAL] How frequently the page is likely to change. This value provides general information to
     * search engines and may not correlate exactly to how often they crawl the page.
     *
     * Valid values are:
     *
     * always
     * hourly
     * daily
     * weekly
     * monthly
     * yearly
     * never
     *
     * The value "always" should be used to describe documents that change each time they are accessed.
     * The value "never" should be used to describe archived URLs.
     *
     * Please note that the value of this tag is considered a hint and not a command. Even though search engine
     * crawlers may consider this information when making decisions, they may crawl pages marked "hourly"
     * less frequently than that, and they may crawl pages marked "yearly" more frequently than that.
     * Crawlers may periodically crawl pages marked "never" so that they can handle unexpected
     * changes to those pages.
     *
     * @url http://www.sitemaps.org/protocol.html#changefreqdef
     * @param string|null $lang language of items content
     * @return string
     */
    public function getSitemapChangefreq($lang = null);

    /**
     * [OPTIONAL] The priority of this URL relative to other URLs on your site. Valid values range from 0.0 to 1.0.
     * This value does not affect how your pages are compared to pages on other sites—it only lets
     * the search engines know which pages you deem most important for the crawlers.
     *
     * The default priority of a page is 0.5.
     *
     * Please note that the priority you assign to a page is not likely to influence the position of your URLs
     * in a search engine's result pages. Search engines may use this information when selecting between URLs
     * on the same site, so you can use this tag to increase the likelihood that your most important pages
     * are present in a search index.
     *
     * Also, please note that assigning a high priority to all of the URLs on your site is not likely to help you.
     * Since the priority is relative, it is only used to select between URLs on your site.
     * @url http://www.sitemaps.org/protocol.html#prioritydef
     *
     * @param string|null $lang language of items content
     * @return string
     */
    public function getSitemapPriority($lang = null);
}


SitemapController.php
<?php

namespace Itstructure\Sitemap;

use Yii;

use yii\console\Controller;

/**
 * Class SitemapController
 * @package Itstructure\Sitemap
 *
 * @property SitemapBuilder $builder
 */
class SitemapController extends Controller
{
    /** @var string Alias to directory contains sitemap-models */
    public $modelsPath = '@console/models/sitemap';

    /** @var string Namespace of sitemap-models files */
    public $modelsNamespace = 'console\models\sitemap';

    /** @var string Path to saving sitemap-files. As webroot: "http://example.com" */
    public $savePathAlias = '@frontend/web';

    /** @var string Name of sitemap-file, saved to webroot: "http://example.com/sitemap.xml" */
    public $sitemapFileName = 'sitemap.xml';

    /** @var array Default config for sitemap builder */
    public $builderConfig = [
        'urlsPerFile' => 100,
    ];

    /** @var string Base Url for sitemap links. As webroot: "http://example.com" */
    public $baseUrl;

    /**
     * @inheritdoc
     */
    public function getHelpSummary()
    {
        return 'Console command for generate sitemap files for Models specified in console config';
    }

    /**
     * Index action for run sitemap creation
     */
    public function actionIndex()
    {
        $dataHandler = new SitemapDataHandler($this->savePathAlias, $this->sitemapFileName, $this->baseUrl, [
            'builderConfig' => $this->builderConfig,
        ]);

        $models = $dataHandler->getModelsClasses($this->modelsPath, $this->modelsNamespace);
        $dataHandler->handleModels($models);
    }
}


interfaces/GoogleImage.php
<?php

namespace Itstructure\Sitemap\interfaces;

/**
 * Interface GoogleImage
 *
 * @url https://support.google.com/webmasters/answer/178636
 *
 * @package Itstructure\Sitemap\interfaces
 */
interface GoogleImage
{
    /**
     * Get list on images assigned to material
     * @param static|self|mixed $material
     * @param string|null $lang Required language of items content
     * @return array
     */
    public function getSitemapMaterialImages($material, $lang = null);

    /**
     * [REQIRED] The URL of the image.
     * In some cases, the image URL may not be on the same domain as your main site. This is fine,
     * as long as both domains are verified in Webmaster Tools. If, for example, you use a content delivery network
     * (CDN) to host your images, make sure that the hosting site is verified in Webmaster Tools OR that you submit
     * your sitemap using robots.txt. In addition, make sure that your robots.txt file doesn’t disallow
     * the crawling of any content you want indexed.
     * @param mixed $image      Image element from [[getSitemapMaterialImages]]
     * @param string|null $lang Required language of item content
     * @return string
     */
    public function getSitemapImageLoc($image, $lang = null);

    /**
     * [OPTIONAL] The geographic location of the image.
     * @example <image:geo_location>Limerick, Ireland</image:geo_location>.
     * @param mixed $image      Image element from [[getSitemapMaterialImages]]
     * @param string|null $lang Required language of item content
     * @return string
     */
    public function getSitemapImageGeoLocation($image, $lang = null);

    /**
     * [OPTIONAL] The caption of the image.
     * @param mixed $image      Image element from [[getSitemapMaterialImages]]
     * @param string|null $lang Required language of item content
     * @return string
     */
    public function getSitemapImageCaption($image, $lang = null);

    /**
     * [OPTIONAL] The title of the image.
     * @param mixed $image      Image element from [[getSitemapMaterialImages]]
     * @param string|null $lang Required language of item content
     * @return string
     */
    public function getSitemapImageTitle($image, $lang = null);

    /**
     * [OPTIONAL] A URL to the license of the image.
     * @param mixed $image      Image element from [[getSitemapMaterialImages]]
     * @param string|null $lang Required language of item content
     * @return string
     */
    public function getSitemapImageLicense($image, $lang = null);
} 

SitemapDataHandler.php
<?php

namespace Itstructure\Sitemap;

use Yii;
use yii\base\BaseObject;
use yii\db\ActiveRecord;
use yii\base\InvalidConfigException;
use yii\helpers\ArrayHelper;
use yii\helpers\Console;
use Itstructure\Sitemap\interfaces\Basic;
use Itstructure\Sitemap\interfaces\GoogleAlternateLang;
use Itstructure\Sitemap\interfaces\GoogleImage;

/**
 * Class SitemapDataHandler
 * @package Itstructure\Sitemap
 *
 * @property SitemapBuilder $builder
 */
class SitemapDataHandler extends BaseObject
{
    /** @var string Path to saving sitemap-files. As webroot: "http://example.com" */
    public $savePathAlias = '@frontend/web';

    /** @var string Name of sitemap-file, saved to webroot: "http://example.com/sitemap.xml" */
    public $sitemapFileName = 'sitemap.xml';

    /** @var array Default config for sitemap builder */
    public $builderConfig = [
        'urlsPerFile' => 10000,
    ];

    /** @var SitemapBuilder Configured builder("file writer") */
    private $_builder;

    /** @var array xml-schemas collected by Model implements */
    private $_schemas = [];

    /** @var string Original site language. Need for swith app languages */
    protected static $_appLanguage;

    /** @var String baseUrl */
    private $baseUrl;

    /**
     * SitemapDataHandler constructor.
     * @param $savePathAlias
     * @param $sitemapFileName
     * @param $baseUrl
     * @param array $config
     */
    public function __construct($savePathAlias, $sitemapFileName, $baseUrl, $config = [])
    {
        $this->savePathAlias = $savePathAlias;
        $this->sitemapFileName = $sitemapFileName;
        $this->baseUrl = $baseUrl;

        parent::__construct($config);
    }

    /**
     * Get list of sitemap model instancies
     *
     * @param string $modelsPath Alias to sitemap-models dir
     * @param string $modelsNamespace
     * @throws \yii\base\InvalidConfigException
     * @return ActiveRecord[]|Basic[]
     */
    public function getModelsClasses($modelsPath, $modelsNamespace)
    {
        $path = Yii::getAlias($modelsPath);
        $files = @scandir($path);

        if ($files === false) {
            throw new InvalidConfigException('Path to sitemap models "' . $modelsPath . '"(' .
                $path . ') is incorrect');
        }

        $buffer = [];
        foreach ($files as $file) {
            if ($file == '.' || $file == '..' || !is_file($path . DIRECTORY_SEPARATOR . $file)) {
                continue;
            }

            // Search only .php files
            $ext = substr($file, -4);
            if ($ext !== '.php') {
                continue;
            }

            // Combine model class name
            $className = $modelsNamespace . '\\' . str_replace($ext, '', $file);

            // Create new model instance
            $model = Yii::createObject($className);

            if (!$model instanceof Basic) {
                $this->printMessage('Warning: model "' . $className . '" does not implement interface ' .
                    '"Itstructure\Sitemap\interfaces\Basic"');

                continue;
            }

            // Add class to list
            $buffer[] = $model;

            // Analysis xml-schemas by $model interfaces
            $this->collectSchemas($model);
        }

        return $buffer;
    }

    /**
     * Start sitemap generate
     * @param Basic[] $models
     */
    public function handleModels(array $models)
    {
        $builder = $this->builder;

        $builder->start();

        foreach ($models as $model) {
            $languages = (array)(isset($model->sitemapLanguages) ? $model->sitemapLanguages : Yii::$app->language);

            if (count($languages) > 1) {
                // Handle separately for each language
                foreach ($languages as $language) {
                    // Swith App language to $language
                    if (isset($model->sitemapSwithLanguages) && $model->sitemapSwithLanguages) {
                        static::setLanguage($language);
                    }

                    $this->handleModel($model, $language);

                    // Restore App language
                    static::restoreLanguage();
                }
            } else {
                // Handle only one language
                $this->handleModel($model, null);
            }
        }

        $builder->finish();
    }

    /**
     * Handle all model content rows
     * @param Basic|ActiveRecord $model
     * @param string|null $lang
     */
    public function handleModel(Basic $model, $lang = null)
    {
        // Handle links to static items
        $items = $model->getSitemapItems($lang);
        if (is_array($items)) {
            // Foreach static model content
            foreach ($items as $item) {
                /* @var $item Basic|GoogleImage|GoogleAlternateLang|ActiveRecord */
                $this->handleItem($item, $lang);
            }
        }

        // Search models in db
        $query = $model->getSitemapItemsQuery($lang);

        if ($query instanceof \yii\db\Query) {
            // Foreach batch models
            $batchSize = static::getModelBatchSize($model);
            foreach ($query->each($batchSize) as $item) {
                /* @var $item Basic|GoogleImage|GoogleAlternateLang|ActiveRecord */
                $this->handleItem($item, $lang);
            }
        }
    }

    /**
     * Handle sitemap item
     * @param Basic|GoogleImage|GoogleAlternateLang|ActiveRecord $item
     * @param string $lang
     * @return bool
     */
    public function handleItem($item, $lang)
    {
        if (is_array($item)) {
            $item = new SitemapArrayItem($item);
            $loc = $item->getSitemapLoc();
            if (empty($loc)) {
                $this->printMessage('Warning: item "' . var_export($item->item, true) .
                    '" does not have a "loc" value');

                return false;
            }
        } elseif (!$item instanceof Basic) {
            $this->printMessage('Warning: item "' . var_export($item, true) .
                '" does not compatible with "Itstructure\Sitemap\interfaces\Basic" or Array');

            return false;
        }

        $builder = $this->builder;
        $url = $builder->newUrl();

        // Basic attributes
        $url->loc($item->getSitemapLoc($lang))
            ->lastmod($item->getSitemapLastmod($lang))
            ->changefreq($item->getSitemapChangefreq($lang))
            ->priority($item->getSitemapPriority($lang));

        // Google image attribute
        if ($item instanceof GoogleImage) {
            $images = $item->getSitemapMaterialImages($item, $lang);

            foreach ($images as $image) {
                $url->addImage(
                    $item->getSitemapImageLoc($image, $lang),
                    $item->getSitemapImageGeoLocation($image, $lang),
                    $item->getSitemapImageCaption($image, $lang),
                    $item->getSitemapImageTitle($image, $lang),
                    $item->getSitemapImageLicense($image, $lang)
                );
            }
        }

        // Google alternate links
        if ($item instanceof GoogleAlternateLang || $item instanceof SitemapArrayItem) {
            $links = $item->getSitemapAlternateLinks();

            foreach ($links as $hreflang => $href) {
                $url->addAlternateLink($hreflang, $href);
            }
        }

        return $builder->writeUrl($url);
    }

    /**
     * Get model sitemapBatchSize attribute
     * @param Basic|ActiveRecord $model
     * @param int $defaultValue
     * @return int
     */
    public static function getModelBatchSize(Basic $model, $defaultValue = 10)
    {
        return isset($model->sitemapBatchSize) ? $model->sitemapBatchSize : $defaultValue;
    }

    /**
     * @return SitemapBuilder
     */
    public function getBuilder()
    {
        if ($this->_builder === null) {
            $config = ArrayHelper::merge([
                'class' => SitemapBuilder::className(),
                'savePathAlias' => $this->savePathAlias,
                'sitemapFileName' => $this->sitemapFileName,
                'schemas' => $this->_schemas,
                'baseUrl' => $this->baseUrl ? $this->baseUrl : Yii::$app->urlManager->baseUrl,
            ], $this->builderConfig);

            $this->_builder = Yii::createObject($config);
        }

        return $this->_builder;
    }

    /**
     * Analysis xml-schemas by $model interfaces
     * @param Basic|GoogleImage|GoogleAlternateLang $model
     */
    public function collectSchemas(Basic $model)
    {
        $this->_schemas['basic'] = 'xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"';

        if ($model instanceof GoogleImage) {
            $this->_schemas['image'] = 'xmlns:image="http://www.google.com/schemas/sitemap-image/1.1"';
        }

        if ($model instanceof GoogleAlternateLang) {
            $this->_schemas['alternate'] = 'xmlns:xhtml="http://www.w3.org/1999/xhtml"';
        }
    }

    /**
     * Show error in console mode
     * @param string $text
     * @param int $fg foreground
     * @param int $bg background
     */
    public function printMessage($text, $fg = Console::FG_YELLOW, $bg = Console::BG_BLACK)
    {
        if (Yii::$app instanceof \yii\console\Application && Yii::$app->controller->interactive) {
            Yii::$app->controller->stdout($text . PHP_EOL, $fg, $bg);
        }
    }

    /**
     * Set new app language and save original language for future restore
     * @param string $lang
     */
    public static function setLanguage($lang)
    {
        static::$_appLanguage = Yii::$app->language;

        Yii::$app->language = $lang;
    }

    /**
     * Restore app language, saved in [[setLanguage]] function
     */
    public static function restoreLanguage()
    {
        if (static::$_appLanguage !== null) {
            Yii::$app->language = static::$_appLanguage;
        }
    }
}
